### ISSUE TITLE
RSDL capabilities syntax proposal -

### ISSUE THREAD

**Issue by chrisspre** [oasis-open/odata-rapid#249]
First draft of a proposal for RSDL capabilities syntax

## Well Known Capabilities

RSDL allows to define structured types and service properties that in combination define which URLs are mapped to which types. It Therfore also implicitly defines which URLs are valid. 
What it doesn't descibe is the capabilities of the service for each of these URLs: which actions, HTTP verbs and query options are valid to be used.

Following proposal is to allow to add a section/block to an RSDL document that maps paths to well-known capabilities.

```RSDL

type Person { 
    key id: string
}

type Company {
    key stockSymbol: string
    employees : [Person]
 }

service { 
    companies: [Company]
    people: [Person]
}

# new section

paths {

    /people : 
    {
        GET { 
            filter
        }
        POST
        PATCH
        DELETE
    }
    /companies/ : {
        GET { 
            filter
            expand { employees { filter }}
        }
        POST
        PATCH
        DELETE
    }
    /companies/{stockSymbol}/people : {
        GET { 
            filter            
        }        
        DELETE
    }
    /companies/{stockSymbol}/people/{id}/$ref : {
        DELETE
        POST
    }
}

```

syntactically this is :

```ABNF

pathsDefinition = "paths" "{" *( pathCapabilities ) "}"


pathCapabilities = path ":" "{" *pathCapability "}"

pathCapability = "GET" [ "{" *getCapabilities "}" ]
               / "PUT"
               / "POST"
               / "PATCH"
               / "DELETE" 

getCapability   = "expand"  ; syntax for nested expand/filter/order tbd 
                / "filter"
                / "ordeby"
                / "top" / "skip" 
                
```

# Capabilities Extensibility 

The above mentioned capabilities are predefined and are scoped to the well known service capabilities of an RSDL service. 
In service implementations  are many ore aspects that go beyond supported HTTP methods and valid query parameters. Complex behaviors, for example throttling or long running transactions, are beyond the scope of what RSDL specifies but would nonetheless fit very well into the RSDL paths.

We are proposing an extensiility where RSDL allows to add so called traits to the RSDL paths definition. These traits are opaque identifiers from the perspective of RSDL but can be translated into actual behavior by the implementing service.

An exmple snippte look like the following
```RSDL

    /competitors/{stockSymbol}: {
        GET {
            expand {
                employees{filter}
            }
            filter
            traits {
                PremiumBusinessModel
                 Throttling { Level=10 }
            }
        }
    }

```

To achive this extensibility we allow to add trait identifiers in the right places of the paths capabilities:

```ABNF


pathCapability = "GET" [ "{" *getCapabilities "}" ]
               / "PUT" [ "{" ?traits "}" ]
               / "POST" [ "{" ?traits "}" ]
               / "PATCH" [ "{" ?traits "}" ]
               / "DELETE" [ "{" ?traits "}" ]

getCapabilities = "expand"  /* tbd */
                / "filter"
                / "ordeby"
                / "top" / "skip"
                / traits

traits = "traits" "{" *trait "}" 

trait = identifier ; tbd: can traits have parameters

```

## open questions

- syntactically: should traits be nested in the `traits` block or should the be in line with the GET, PUT, ... capabilities. (if the later we might need to quote traits or have a prefix to visually distinguish them)
- syntactically: nesting of "block's" (sections enclosed in "{" "}") is becoming quite deep, and there is no obvious pattern when a ":" is needed and when not.
- should traits be just plain identifiers or can the be parameterized (e.g. `serviceLevel(level=premium)`  or syntactically similar but always identifier + key/value pairs)
-

**Comment by ralfhandl** [comment]
Wondering about `PATCH` and `DELETE` here:

> /people : 
>     {
>         GET { 
>             filter
>         }
>         POST
>         PATCH
>         DELETE

Is `/people` including key access? Or is key access a second path `/people/{id}`?

Why a trailing slash here?

>     /companies/ : {
>         GET { 
>             filter
>             expand { employees { filter }}
>         }

**Comment by chrisspre** [comment]
> Wondering about `PATCH` and `DELETE` here:

> Is `/people` including key access? Or is key access a second path `/people/{id}`?

Good point. This should be split one path /people, one path /people/{id}/

> Why a trailing slash here?

Just copy paste error (which we had a parser/editor for this :-) )

**Comment by chrisspre** [comment]
`GET   {}` => means no query options are supported, not even $select
`GET  `  => means the _default_ query options are supported

what does "default" mean?
Idea:   service RSDL has a service level default 

============

`expand ` =>  can expand every property
`expand { } ` => means none can be expanded (sort of useless ; equivalent to not mentioning `expand`
`expand { <property1> } ` => expand only for the given property1 and no nested 

`expand { <property1> } ` => expand only for the given property1

**Comment by chrisspre** [comment]
`GET   {}` => means no query options are supported, not even $select
`GET  `  => means the _default_ query options are supported

what does "default" mean?
Idea:   service RSDL has a service level default 

============

`expand ` =>  can expand every property
`expand { } ` => means none can be expanded (sort of useless ; equivalent to not mentioning `expand`
`expand { <property } ` => expand only for the given property

**Comment by chrisspre** [comment]
TODO: how are custom query options expressed in RSDL (and if.

**Comment by mikepizzo** [comment]
I wonder if, for services that are recursive, if we can have a way to annotate the canonical path to say "any route to this item has the given set of properties"

for example:

~~~
/people* : 
   {
    GET { 
        filter
    }
~~~

the * could mean that any route to a person from the /people entity set supports GET with a filter.

**Comment by marabooy** [comment]
@chrisspre & @mikepizzo The current specification kind of touches on what the allowed list will be in terms of properties 
```
expand { <property1> }
```
Would it be possible to have a way to describe the denied properties in here? Something along the lines of 
```
expand { <!property1> } => // which would allow all properties which are navigation properties but not on property one. (not sure what the negation in RSDL is supposed to look like using the exclamation mark to showcase this.)
```
This could be a great shorthand for large services which may need to only deny a small set of properties from being selected, expanded or filtered on whilst also providing a good bridge with the capabilities vocabulary as well.

**Comment by chrisspre** [comment]
> describe the denied properties in here? Something along the



> This could be a great shorthand for large services which may need to only deny a small set of properties from being selected, expanded or filtered on whilst also providing a good bridge with the capabilities vocabulary as well.

David, yes, I think the ability to specify which properties are to be expected to be returned would be good. But i woudl want to make this a bit more structured, avoid the negation and not add it to expand but rather select.

My suggestion woudl be 

``` 
select {
    readonly { id, lastUpdateDateTime }
    writeonly { password, largeBlobOfData }
}
```
with the meaning that 
- the specified writeonly (structural) properties are not returned in a response
- the specified readonly (structural) properties can not be specified in the request. 

Need to think about how this works together with already specified HTTP method and how defaults for this would work.

**Comment by garethj-msft** [comment]
> First draft of a proposal for RSDL capabilities syntax
> 
> ## Well Known Capabilities
> RSDL allows to define structured types and service properties that in combination define which URLs are mapped to which types. It therfore also implicitly defines which URLs are valid.
> What it doesn't describe is the capabilities of the service for each of these URLs: which actions, HTTP verbs and query options are valid to be used.
> 
> Following proposal is to allow to add a section/block to an RSDL document that maps paths to well-known capabilities.
> 
> ```
> 
> type Person { 
>     key id: string
> }
> 
> type Company {
>     key stockSymbol: string
>     employees : [Person]
>  }
> 
> service { 
>     companies: [Company]
>     people: [Person]
> }
> 
> # new section
> 
> paths {
> 
>     /people :  {
>         GET { 
>             filter
>         }
>         POST
>     }
>     /people/{id} : {
>         GET 
>         DELETE
>         PATCH
>     }
>     /companies : {
>         GET { 
>             filter
>             expand { employees { filter }}
>         }
>         POST
>         PATCH
>         DELETE
>     }


I still contend that the paths should not only be about capabilities, but also redundant type bindings.
A key complain we have is that in order to work out what the type of a path in OData is, I have to go follow the chain of nav properties.   

We should be solving for that complaint IMO, even at the cost of breaking DRY.

To me, as someone who will always go directly to the paths section first once we have this, it is really strange to have to then go to the service and frequently a long chain of type definitions to work out what the type of the response for the path is.

**Comment by ralfhandl** [comment]
> A key complain we have is that in order to work out what the type of a path in OData is, I have to go follow the chain of nav properties.

I think this is a tooling issue. For example in a VS Code extension for RSDL I would expect to get the type information hovering over a path or path segment.

**Comment by ralfhandl** [comment]
> I wonder if, for services that are recursive, if we can have a way to annotate the canonical path to say "any route to this item has the given set of properties"

I like the idea and wonder about the syntax. Would it be
* `path to /people` meaning any path that arrives at `/people`? 

Which raises the question: 
* How do we express navigation property bindings?

**Comment by marabooy** [comment]
> > describe the denied properties in here? Something along the
> 
> > This could be a great shorthand for large services which may need to only deny a small set of properties from being selected, expanded or filtered on whilst also providing a good bridge with the capabilities vocabulary as well.
> 
> David, yes, I think the ability to specify which properties are to be expected to be returned would be good. But i woudl want to make this a bit more structured, avoid the negation and not add it to expand but rather select.
> 
> My suggestion woudl be
> 
> ```
> select {
>     readonly { id, lastUpdateDateTime }
>     writeonly { password, largeBlobOfData }
> }
> ```
> 
> with the meaning that
> 
> * the specified writeonly (structural) properties are not returned in a response
> * the specified readonly (structural) properties can not be specified in the request.
> 
> Need to think about how this works together with already specified HTTP method and how defaults for this would work.

I guess this could also be applicable to filter as well.

**Comment by chrisspre** [comment]
> 
> I think this is a tooling issue. For example in a VS Code extension for RSDL I would expect to get the type information hovering over a path or path segment, and Control-Click to jump to that type.

I agree that tooling helps tremendously with that and that is very useful for the author of the model. 
But a causal reader might not have that information readily available. Especially in a phase where the mental model of the types and their relationship is formed. The names of the property at the end of the path helps but it can't always be the type's name.

I am thinking that the following might be helpful:

```
path /orders/{id}/items  type  OrderItem {
    GET { ... }
}
```
(not sure about the keyword `type` but something along these lines)
For the original author it is obvious and easy to add the type. The parser/translator/service need to check the type anyways since the properties used in the capabilities (e.g. filter capability refers to property names) need to be checked by the and they are found on the type.  And for the casual reader it is right there.

**Comment by ralfhandl** [comment]
> But a causal reader might not have that information readily available.

The casual reader might prefer a nicely formatted representation, with a graph of the service's model. We should be able to produce nice markdown from RSDL / CSDL.

> For the original author it is obvious and easy to add the type. 

Nope, it violates the DRY principle and requires "double-entry book keeping" between `path` and `type` constructs, and

> Especially in a phase where the mental model of the types and their relationship is formed. 

As a service author I don't want to waste time on this synchronization.

I think we have several competing goals here:
  1. Is RSDL a 
    a. service _description_ language targeting _readers_, or a 
    b. service _definition_ language targeting _authors_?
  2. Is RSDL targeting 
    a. _model first_ and adding path-based restrictions later, or
    b. _path first_ with an implicit model (or no consistent model at all, as in OpenAPI)?

**Comment by chrisspre** [comment]
> Nope, it violates the DRY principle and requires "double-entry book keeping" between `path` and `type` constructs, which is painful

I would argue it doesn't (debating myself though). 
Writing  the name of the type is similar to declaring a type of a variable when the right hand side clearly indicated the type in `int x = 1; ` the int is technically redundant (in most cases).

**Comment by ralfhandl** [comment]
> Writing the name of the type is similar to declaring a type of a variable

The difference is that variable names can be chosen freely, but paths are derived from the model. If I have 
~~~rsdl
service {
  foo: [Foo]
}

type Foo {
   key id: String
   bar: Bar
}
~~~
then I can't decide to also have
~~~rsdl
path /foo/{id}/bar type Qux
~~~
The model already says that the type of this path is `Bar`.

Let's decide on the goals we want to achieve with RSDL:

> * Is RSDL a
>   a. service _description_ language targeting _readers_, or a
>   b. service _definition_ language targeting _authors_?
> * Is RSDL targeting
>   a. _model first_ and adding path-based restrictions later, or
>   b. _path first_ with an implicit model (or no consistent model at all, as in OpenAPI)?

**Comment by chrisspre** [comment]
> Let's decide on the goals we want to achieve with RSDL:
> 
> > * Is RSDL a
> >   a. service _description_ language targeting _readers_, or a
> >   b. service _definition_ language targeting _authors_?
> > * Is RSDL targeting
> >   a. _model first_ and adding path-based restrictions later, or
> >   b. _path first_ with an implicit model (or no consistent model at all, as in OpenAPI)?

I think we are in agreement that it is b) (language purpose) for the first question and a (model first) to the second. 
But we are having authors that are not spending a major amount of time on the service definition and change it as a team effort. The additional information of the type after the path would help them substantially. 

An alternative could be to nest the path declaration under the corresponding type.

**Comment by ralfhandl** [comment]
We could make the type after the path optional:
- If you write a type there, you will get a compiler error if you guessed the wrong type
- If you don't write a type, you are on the safe side
